### PR TITLE
feat: Add contrib module

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

adds contrib directory and seeds it with a poi scan plot. the idea behind `contrib` is to have often-repeated code e.g. in the example noteboook centrally managed, however nothing in the non-`contrib` part of pyhf  (src/pyhf) must depend on on it.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add pyhf.contrib
* Add pyhf.contrib.viz.brazil for POI scan "Brazil band" plots
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
While doing testing of visualization libraries is hard, it is possible. Can we get tests for `contrib/viz`?

**Comment by lukasheinrich** [comment]
@matthewfeickert not sure.. any ideas?

**Comment by lukasheinrich** [comment]
https://towardsdatascience.com/unit-testing-python-data-visualizations-18e0250430 ?

**Comment by kratsg** [comment]
> @matthewfeickert not sure.. any ideas?

https://matplotlib.org/3.1.1/devel/testing.html#writing-an-image-comparison-test

Kinda works like betamax/vcr works with python requests... record and compare existing recording.

**Comment by kratsg** [comment]
So one thing I was playing around with locally for `contrib` was using it to attach new functions to the existing classes we have. E.G. when importing `contrib`, it would automatically add a `parameter_scan` function like so

in `contrib/__init__.py`
```python
from ..pdf import Model
from .utils import parameter_scan
Model.parameter_scan = parameter_scan
```

since I kinda treated the `contrib` as a "decoration" of the existing code, rather than adding new things...

**Comment by lukasheinrich** [comment]
monkey patching the rest of the codebase seems a bit too magical to me :).

**Comment by matthewfeickert** [comment]
> @matthewfeickert not sure.. any ideas?

I think we can pass on having them in this PR and then make an Issue to get them in soon.

File: src/pyhf/contrib/viz/brazil.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import numpy as np
+
+
+def plot_results(ax, mutests, tests, test_size=0.05):
+    cls_obs = np.array([test[0] for test in tests]).flatten()
+    cls_exp = [np.array([test[1][i] for test in tests]).flatten() for i in range(5)]
+    ax.plot(mutests, cls_obs, c='k')
```
Comment (matthewfeickert) [inline src/pyhf/contrib/viz/brazil.py]: I'd suggest using verbose color names
```suggestion
    ax.plot(mutests, cls_obs, c='black')
```

File: src/pyhf/contrib/viz/brazil.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    for i, c in zip(range(5), ['k', 'k', 'k', 'k', 'k']):
+        ax.plot(mutests, cls_exp[i], c=c, linestyle='dotted' if i != 2 else 'dashed')
+    ax.fill_between(mutests, cls_exp[0], cls_exp[-1], facecolor='y')
```
Comment (matthewfeickert) [inline src/pyhf/contrib/viz/brazil.py]:
```suggestion
    ax.fill_between(mutests, cls_exp[0], cls_exp[-1], facecolor='yellow')
```

File: src/pyhf/contrib/viz/brazil.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    ax.fill_between(mutests, cls_exp[1], cls_exp[-2], facecolor='g')
```
Comment (matthewfeickert) [inline src/pyhf/contrib/viz/brazil.py]:
```suggestion
    ax.fill_between(mutests, cls_exp[1], cls_exp[-2], facecolor='green')
```

File: src/pyhf/contrib/viz/brazil.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    ax.plot(mutests, [test_size] * len(mutests), c='r')
```
Comment (matthewfeickert) [inline src/pyhf/contrib/viz/brazil.py]:
```suggestion
    ax.plot(mutests, [test_size] * len(mutests), c='red')
```

**Comment by lukasheinrich** [comment]
I settled for https://github.com/matplotlib/pytest-mpl which seems to work

**Comment by lukasheinrich** [comment]
I tested this with an intentionally failing build. The reference images are in `tests/contrib/baseline`

screenshot

maybe in a new PR we could add some github CI magic that would *post* the reference, new and diff image to the PR when a visual test fails.

**Comment by kratsg** [comment]
> maybe in a new PR we could add some github CI magic that would _post_ the reference, new and diff image to the PR when a visual test fails.

I like this idea. It should be doable with the GitHub API + octokit.

**Comment by matthewfeickert** [comment]
Why is the coverage dropping so sharply?

> Merging #694 into master will decrease coverage by 70.96%.

This seems wrong, but I'm confused as to what is happening.

**Comment by lukasheinrich** [comment]
@matthewfeickert I think the comment  is a old run of codecoov, the latest check on codecov should be green

**Comment by matthewfeickert** [comment]
> @matthewfeickert I think the comment is a old run of codecoov, the latest check on codecov should be green

I rebased the PR to trigger a new run of codecov. Hopefully once the CI runs all the way through it will change. If so, I'll merge in when back from dinner tonight (or one of you can also). :)

**Comment by matthewfeickert** [comment]
Nope. Sadly Codecov is just being weird and is reporting a coverage of 21% now. WTF. :(

**Comment by matthewfeickert** [comment]
I've reported this now: https://community.codecov.io/t/large-coverage-drop-after-addition-of-100-covered-file/842

**Pull request merged**